### ISSUE TITLE
MacOS standard "Window" menu is in the wrong place and sometimes moves around

### ISSUE THREAD

**Issue by nobugshere** [wxWidgets/wxWidgets#22870]
### Description

MacOS apps have a standard "Window" menu item, whether you want it or not :-).  By convention, it is placed to the left of the "Help" menu but, in apps using wxWidgits, it appears to the right (see first screenshot of the Hello World app). 

![wxDefault](https://user-images.githubusercontent.com/61946508/195725017-34da8cdc-2c20-46f0-b292-f884aac3e1e2.png)

In a more complicated app, the "Window" menu starts off on the right but jumps to the left of the "File" menu when the window is minimized then restored.  I haven't been able to reproduce that behavior in a simple test case.

If I insert code to add the "Window" menu right before adding the "Help" menu it seems to place it where I want it, and it gets the standard system entries.

![Hack](https://user-images.githubusercontent.com/61946508/195725448-8f6c7725-1fdc-400c-90e3-d8cc0b903b75.png)

However, in the complicated app, this trick works at first, but after a minimize/restore cycle, the menu stays in the right place but stops working.

It seems like this is something that should be handled by the framework unless the user needs to customize it.

### Platform and version information

- wxWidgets version  you use:
3.2.1 (installed via Homebrew)
- wxWidgets port  you use:
wxOSX
- OS  and its version:
M1 Max macOS Monterey 12.6 and Xcode 14.0.1

**Comment by csomor** [comment]
Hi

You could turn the auto-window-menu off using `wxMenuBar::SetAutoWindowMenu(false)` but since all the standard window manipulation commands are in there, this should rarely be used ...

Right now I'm not reordering the menus in a menubar, only adding the apple menu in the front and the window menu at the end, so the recommended way to arrive at the correct order right now would be, e.g. if you'd take the minimal sample to add a dummy window menu

```
menuBar->Append(fileMenu, "&File");
menuBar->Append(new wxMenu(), "&Window");
menuBar->Append(helpMenu, "&Help");
```

I could create a PR which does an autoreorder, but I'm a little bit relucatant to do that

**Comment by vadz** [comment]
I think we need to add "Window" menu at the end except if the last menu is "Help", in which case we should put it before it.

But it's not totally clear to me when does the problem arise currently, does it happen in any of our samples? If not, could you please provide a [minimal patch](https://www.wxwidgets.org/develop/how-to-submit-patches/) allowing to reproduce it?

TIA!

**Comment by csomor** [comment]
> But it's not totally clear to me when does the problem arise currently, does it happen in any of our samples? If not, could you please provide a [minimal patch](https://www.wxwidgets.org/develop/how-to-submit-patches/) allowing to reproduce it?

Minimal Sample already shows this, as it has not built-in window menu, but a built-in help menu

**Comment by csomor** [comment]
could you please check with PR #22871

**Comment by nobugshere** [comment]
> Minimal Sample already shows this, as it has not built-in window menu, but a built-in help menu

Yes, just to be clear, the first screenshot is from the "Hello World" tutorial [here](https://docs.wxwidgets.org/latest/overview_helloworld.html).  The second one is after adding essentially the same sequence @csomor suggested.

About the other problem mentioned, where the Window menu jumps to the left end or stops working after minimize/restore, I have not been able to reproduce it in a simple test case.  I'll keep looking at the original app to see if its doing something strange.

Thank you.

**Comment by nobugshere** [comment]
> could you please check with PR #22871

@csomor, please, excuse the noob question, do you have a build you want me to download, or should I clone you're repo and build it myself?  So far, I've only worked with the binary build installed via Homebrew.  I'm still wrapping my head around the GH way :-)

**Comment by csomor** [comment]
@nobugshere Hi Richard, yes, clone the PR Branch and build locally. If you'd use the github CLI this would be 

```
gh repo clone wxWidgets/wxWidgets
cd wxWidgets
gh pr checkout 22871 --recurse-submodules
```

I'd then do a configure with a custom location install using `--prefix` option, so that things stay isolated

**Comment by nobugshere** [comment]
Hi Stefan,  @csomor
Thanks for the tips.  I was able to build your branch and test the patch.  But, I also figured out why the Window menu would sometimes jump to the left end of the menu bar in an actual app.  On minimize/restore, the app would remove all of the menu items from the menu bar and add them all back.  (Apparently it's the same routine that would be called if the user selects a different language.)  To do this it calls GetMenuCount() and Remove() in a loop.  However, the count that is returned does not seem to include the default "Window" menu.  [PR #22878](https://github.com/wxWidgets/wxWidgets/pull/22878#issue-1410352563) is a test case that models the behavior of the app.  The "Bogus" menu has three items:  display the return value of GetMenuCount; remove the menu bar items; remove and re-add the menu bar items.  It appears there is an internal list that the "Window" menu is not being added to.

I hope this helps.  Thanks again!

**Comment by csomor** [comment]
@nobugshere Thanks for the sample Richard. Yes, I feared about internal and external menu lists getting out of synch, when manipulated directly - and not having a window menu on their own. I'll try to find the best solution.

**Comment by csomor** [comment]
@nobugshere in your app - why is creating a new `wxMenuBar` with the new language and setting this as the new menubar onto the frame not an option ?

**Comment by nobugshere** [comment]
@csomor, The honest answer is: I don't know.  I had the same thought but I didn't write the app. The app in question is [pwsafe](https://github.com/pwsafe/pwsafe).  I used it on Windows for many years.  I recently switched to a Mac and discovered it had a Mac port.  So, I got interested in fixing a few minor issues in it, mostly as an exercise to learn Mac/Xcode/Git/GH/... Etc.  

Question: If I delete the old wxMenuBar object, would that leak the wxMenu object for the lingering Window menu, or would the destructor clean that up for me?
